docs/fundamentals/code-analysis/quality-rules/ca1844.md

```diff
@@ -0,0 +1,48 @@
+---
+title: "CA1844: Provide memory-based overrides of async methods when subclassing 'Stream' (code analysis)"
+description: "Learn about code analysis rule CA1844: Provide memory-based overrides of async methods when subclassing 'Stream'"
+ms.date: 05/18/2021
+ms.topic: reference
+f1_keywords:
+- CA1844
+- ProvideStreamMemoryBasedAsyncOverrides
+helpviewer_keywords:
+- ProvideStreamMemoryBasedAsyncOverrides
+- CA1844
+author: NewellClark
+dev_langs:
+- CSharp
+- VB
+---
+# CA1844: Provide memory-based overrides of async methods when subclassing 'Stream'
+
+| | Value |
+|-|-|
+| **Rule ID** |CA1844|
+| **Category** |[Performance](performance-warnings.md)|
+| **Fix is breaking or non-breaking** |Non-breaking|
+
+## Cause
+
+A type derived from <xref:System.IO.Stream> overrides <xref:System.IO.Stream.ReadAsync(System.Byte[],System.Int32,System.Int32,System.Threading.CancellationToken)> or <xref:System.IO.Stream.WriteAsync(System.Byte[],System.Int32,System.Int32,System.Threading.CancellationToken)> but does not override <xref:System.IO.Stream.ReadAsync(System.Memory{System.Byte},System.Threading.CancellationToken)> or <xref:System.IO.Stream.WriteAsync(System.ReadOnlyMemory{System.Byte},System.Threading.CancellationToken)>, respectively.
+
+## Rule description
+
+The memory-based methods were added to improve performance, which they accomplish in multiple ways:
+
+1. They return `ValueTask` and `ValueTask<int>` instead of `Task` and `Task<int>`, respectively, and
+2. They allow any type of buffer to be passed in without having to perform an extra copy to an array.
+
+In order to realize these performance benefits, types that subclass <xref:System.IO.Stream> must provide their own memory-based implementation. Otherwise, the default implementation will be forced to copy the memory into an array in order to call the array-based implementation, resulting in reduced performance, particularly when the caller passes in a <xref:System.Memory%601> or <xref:System.ReadOnlyMemory%601> instance that is not backed by an array.
```

Comment:
```suggestion
In order to realize these performance benefits, types that derive from <xref:System.IO.Stream> must provide their own memory-based implementation. Otherwise, the default implementation will be forced to copy the memory into an array in order to call the array-based implementation, resulting in reduced performance. When the caller passes in a <xref:System.Memory%601> or <xref:System.ReadOnlyMemory%601> instance that's not backed by an array, performance is affected more.
```